src/openrct2/core/FileIndex.hpp

```diff
@@ -0,0 +1,286 @@
+
+        auto endTime = std::chrono::high_resolution_clock::now();
+        auto duration = (std::chrono::duration<float>)(endTime - startTime);
+        Console::WriteLine("Finished building %s in %.2f seconds.", _name.c_str(), duration.count());
+        return items;
+    }
+
+    std::tuple<bool, std::vector<TItem>> ReadIndexFile(const DirectoryStats &stats) const
+    {
+        bool loadedItems = false;
+        std::vector<TItem> items;
+        try
+        {
+            log_verbose("FileIndex:Loading index: '%s'", _indexPath.c_str());
+            auto fs = FileStream(_indexPath, FILE_MODE_OPEN);
+
+            // Read header, check if we need to re-scan
+            auto header = fs.ReadValue<FileIndexHeader>();
+            if (header.MagicNumber == _magicNumber &&
+                header.VersionA == FILE_INDEX_VERSION &&
+                header.VersionB == _version &&
+                header.LanguageId == gCurrentLanguage &&
+                header.Stats.TotalFiles == stats.TotalFiles &&
+                header.Stats.TotalFileSize == stats.TotalFileSize &&
+                header.Stats.FileDateModifiedChecksum == stats.FileDateModifiedChecksum &&
+                header.Stats.PathChecksum == stats.PathChecksum)
+            {
+                // Directory is the same, just read the saved items
+                for (uint32 i = 0; i < header.NumItems; i++)
+                {
+                    auto item = Deserialise(&fs);
+                    items.push_back(item);
+                }
+                loadedItems = true;
+            }
+            else
+            {
+                Console::WriteLine("%s out of date", _name.c_str());
+            }
+        }
+        catch (const Exception &)
+        {
+            Console::Error::WriteLine("Unable to load index: '%s'.", _indexPath.c_str());
+        }
+        return std::make_tuple(loadedItems, items);
+    }
+
+    void WriteIndexFile(const DirectoryStats &stats, const std::vector<TItem> &items) const
+    {
+        try
+        {
+            log_verbose("FileIndex:Writing index: '%s'", _indexPath.c_str());
+            auto fs = FileStream(_indexPath, FILE_MODE_WRITE);
+    
+            // Write header
+            FileIndexHeader header = { 0 };
+            header.MagicNumber = _magicNumber;
+            header.VersionA = FILE_INDEX_VERSION;
+            header.VersionB = _version;
+            header.LanguageId = gCurrentLanguage;
+            header.Stats = stats;
+            header.NumItems = (uint32)items.size();
+            fs.WriteValue(header);
+    
+            // Write items
+            for (const auto item : items)
+            {
+                Serialise(&fs, item);
+            }
+        }
+        catch (const Exception &)
+        {
+            Console::Error::WriteLine("Unable to save index: '%s'.", _indexPath.c_str());
+        }
+    }
+
+    static uint32 GetPathChecksum(const std::string &path)
```

Comment:
I think its better to move to FNV-1a for string hashing.

Reply:
This was the method used in vanilla. You are welcome to change it, I certainly don't want to spend any time on changing this when its only a quick naïve method to see if files have been changed (even those their dates haven't) which is extremely unlikely.

Reply:
I suppose, well it would be better to move all hashing/string utils into its own file anyway, i'll keep that in mind for later.